Fix #1419: replace regex implementation with scala-native/re2s

LeeTibbert opened the pull request:
#### Partial TO DO List:

- [X] Update LICENSE.TXT.

- [ ]  Figure out Travis CI issues with too_many/to_large unit-test Suites.

- [X]  Remove `breakable` idiom from Machine.scala.

- [X]  Check that `breakable` idiom in on-demand-only ExecTest.scala.disabled  is acceptable.

- [X] Are the Unicode tables latest Unicode 12.0 (March 2019)? They are probably two or more
       generations older. Deferred to Issue #1542.

- [X] Can Unicode tables be shared with j.l.Character.scala for reduced binary size, consistency,
       and ease of updating the Unicode tables. Deferred to Issue #1542 

- [X] Fix the PatternSuite quoted string defect reported in Issue #1504 

- [X] Remove cre2 implementation if favor of re2s.

- [X] Consider compressing the Unicode tables and expand at first use. Do not introduce
      dependencies such as xz or zstd. Deferred to Issue #1542 

- [X] Implement the two routines in re2s which currently have stub annotations because they are
      `???`.

- [ ]  Create Issue about missing region support in regex.Matcher.scala. That can be a subtask
        and need not hold up this main PR.

- [ ]  Explore implementing  Least Recently Used cache for patterns in regex.Pattern.scala. Consider
     carefully the interaction with the caching in re2s.RE2.scala.

- [ ]  Do a better job inuser facing documentation file  javalib.rst of describing how re2s
        implementation differs from j.u.regex. Cross check with Issues in original re2s Repository. 

##### ========================================================================
  * This is work in progress.  The porting work is my original
    work but it is based on earlier work at
    [re2s](https://github.com/scala-native/re2s).    See next bullet item.

    It is still pretty raw, but I present it for high level & conceptual
    review, especially of the proposed file location and package names.

    My thanks & appreciation to the community for comments & constructive
    evolution.

  * I want/need to send a note to the original scala-native/re2s
    to thank them for their work.

  * No effort was made in this PR to remove the current re2 implementation.
    That can be a follow on PR once this code has been merged and
    exercised.

  * Issue #1418 requested "Support for scala-native/re2s

    That issue is now fixed and relevant unit-tests provided.

  * Key to test suites:

      - Any unit test marked re2s.*Suite.scala runs and passes in Travis. 

      -   Any test marked *Suite.scala.disabled runs & passes in my private environment. 
        Recently I have been encountering problems with Travis when presenting
        it with too many Suites, Issue #1528 .  All *.disabled Suites, except TestExec.scala, are intended
        to be enabled as soon as Travis can handle them 
 
      - ExecTest.scala.disabled is intended as an "on-demand" test. That is, it resides in the
      repository and can be found and enabled by developers but it is not run on each CI.
      The test is exhaustive and takes tens of minutes to run.  Essential for private development,
      not suitable for each CI run.
 
Documentation:

  * To Be Specified before graduating from WIP.

  * Of course, any release note needs to copy the standard and
    obligatory warning that re2 approaches but is not
    java.mumble.Regex.

Testing:

  * All tests, including ExecTest, pass in my private development environment.
    This puts SN re2s at parity with JVM re2s.   

    Some of the JVM re2s test suites have commented out code and there are, I believe,
    7 Issues in the JVM re2s showing cases where re2s is not at parity with its code base.
    The code base seems to be a mixture of original go re2s, possibly the c port of re2,
    and references to the original go code.  

  * All tests excluding ExecTest pass on Travis but only N  -2 can pass at any one time,
     Issue #1528. Same caveats on test suite parity as above.

LeeTibbert commented:
This PR is now ready for detailed review. 

After the cited PRs have been merged, I will kick off a final build of this PR, to ensure it is ready
for merge.

The last significant piece of work remaining is for me to bring the original PR description up to date.
I want to do that when I am rested & my humor has improved.

densh commented:
@LeeTibbert All of the dependent PRs apart from #1497 are merged. The last remaining one needs a rebase/reformat and it should be good to go as well.

densh commented on nativelib/src/main/scala/scala/re2s/UnicodeTables.scala:
```diff
@@ -0,0 +1,2483 @@
+  final val Tifinagh               = _Tifinagh
+  final val Ugaritic               = _Ugaritic
+  final val Vai                    = _Vai
+  final val Yi                     = _Yi
```
These `final val`s which just alias other `final val`s seem to be unnecessary. Why not just make the original public?

densh commented on nativelib/src/main/scala/scala/re2s/Compiler.scala:
```diff
@@ -0,0 +1,270 @@
+// Original Go source here:
+// http://code.google.com/p/go/source/browse/src/pkg/regexp/syntax/compile.go
+
+package scala.re2s
```
`scala.scalanative.re2s` is a better package name here. We can't jus tadd things to `scala` namespace.

LeeTibbert replied:
Changed as suggested. Builds in Travis. 
Thank you.  By the light of day, "What was I thinking?"

This is _exactly_ the kind of thing I was trying to detect in this PR.

densh commented on unit-tests/src/test/scala/re2s/RE2PatternSuite.scala:
```diff
@@ -0,0 +1,210 @@
+
+  private def reserialize(obj: Pattern) = {
+
+    obj // Remove when Object*Stream methods are implemented
```
We don't really plan to support `Object*Stream` (i.e., java serialization), so might as well just drop code that depends on it.

LeeTibbert replied:
Changed as recommended.  Thank you for the improvement & simplification.  Less lore to
remember or re-create.

Passes Travis CI.

LeeTibbert commented on nativelib/src/main/scala/scala/re2s/Machine.scala:
```diff
@@ -0,0 +1,400 @@
+      flag = in.context(pos)
+    }
+
+    breakable {
```
I will look into this.  I had been treating the re2s code as a black box and focusing on
ensuring the output passes tests.  Looks like I have to level up and look inside ;-)

I took one run at this and will do another when my mind is clear.  The code is _ugly_
in the max. It does breaks inside of if statement blocks, with no else clause. So the 
now-known-to-be-brainless edits did on my first pass blew up and cost me half a day.
The excuse "It made sense at the time" never worked in my (extended) childhood
and it does not seem to be working now.

I need to trace the code carefully but this looks like some kind of state machine distorted
by successive language translations. It worked for the translators and I am using
their work, so I am not complaining, just trying to figure things out.

I may look at the original Go code to see the ancestor. Perhaps there is some
clarity there.   

I personally am of the school that Exceptions should be used for exceptional cases, not
normal flow control.

LeeTibbert replied:
Prompted by your comment, I grep`d all other re2s source and unit-test files.  The .scala
files had the one `breakable` you caught.  


There is one unit-test with a `breakable`.  It is not performance critical but it is on my
list to fix also.

LeeTibbert replied:
As recommended, I removed the `breakable` idiom from Machine.match_().  That is a
key piece of re2s, so I think all of re2s will benefit from the now faster code.

Travis CI is all green.

densh replied:
Looks much better now! 👍

LeeTibbert replied:
Thank you.  Can I ask for some guidance, if you please?  

ExecTest.scala is the only other place now in the SN re2s codebase which uses
`breakable`.  There are two instances, in two logically distinct places.  At runtime this
results in, I think, six executions. The same transform of the `breakable` to a 
subroutine/method can be done but may not be economical. 

ExecTest.scala is a test suite (yes, naming convention inconsistency, jury still out on that one)
intended to be "on-demand" and not done every CI run. It is an exhaustive test which takes
tens of minutes (20? 30?  I walk away and go to lunch.)  It should be available in the repository
so that developers of SN itself can find and use it in their private directories.  It gives 
comfort when one knows that it runs successfully.

Expunge `breakable` from the code base has benefit but is saving six executions over
the span of 20 or more minutes on a test run once every human lunar landing worth
the devo time?

If yes, I will proceed when I come up for air from Unicode.   A long read for a one word answer ;-)

densh replied:
It's perfectly fine to keep `breakable` in tests. My comment above only concerns non-test code.

LeeTibbert replied:
Understood.  Thank you for the guidance.

LeeTibbert commented on nativelib/src/main/scala/scala/re2s/UnicodeTableData.scala:
```diff
@@ -0,0 +1,2572 @@
+trait UnicodeTableData {
+  private val _Lm: Array[Array[Int]] =
+    Array(
+      Array(0x02b0, 0x02c1, 1),
```
Understood, time to open the black box and apply what you have taught me (my words, not yours).  This is pending depending upon what I can figure out about common data with
Character.scala.

densh replied:
Another option here is to store in "column major format" if it makes the code any simpler/easier to read than computing indexes into a giant single dimensional array that contains all the data. For example, instead of:
```
val myarr = Array(
  Array(a1, b1, c1),
  Array(a2, b2, c2),
  ...
```
We can do:
```
val myarrA = Array(
  a1,
  a2,
  ...
)
val myarrB = Array(
  b1,
  b2
  ...
)
val myarrC = Array(
  c1,
  c2
  ...
)
```

LeeTibbert replied:
Good suggestion. Parallel arrays make the code easier to both to write and get correct
and to read later on.

I am currently working on ripping re2c out.  With your agreement, I'll work on this
after rec2 has seen the door.

ekrich replied:
This is the approach I took based on @sjrd 's help for upper and lower case. The range is searched via [binary search here](https://github.com/scala-native/scala-native/blob/master/javalib/src/main/scala/java/lang/Character.scala#L1455) and then the other arrays are to calculate offsets etc. as explained starting [here.](https://github.com/scala-native/scala-native/blob/master/javalib/src/main/scala/java/lang/Character.scala#L1215)

I reviewed `re2s` a bit and see they are doing something similar although I would need to study harder to really understand exactly what they are doing. The nested arrays are an interesting approach but I'm not sure why that approach was taken. It would be great to unify the usage of unicode as then we code generate new arrays based on a new unicode version. Sharing the data in runtime is a really good idea as well.

densh replied:
@ekrich It really seems like unicode unification should be a stand-alone PR. It doesn't seem to be trivial and it distracts from the main goal of the current PR: getting rid of old regex implementation.

densh replied:
@LeeTibbert Yes, lets focus on getting rid off c-based re2 in this PR.

ekrich replied:
@densh and @LeeTibbert I totally agree, just wanted to affirm my support for your ideas.

LeeTibbert commented on nativelib/src/main/scala/scala/re2s/UnicodeTableData.scala:
```diff
@@ -0,0 +1,2572 @@
+package scala.re2s
+
+trait UnicodeTableData {
```
I understand & agree.  I think you have read and support my concerns with "Single Points of
Truth"

I will check.  I did a first pass and nothing obvious fell out.

After I get `breakable` resolved and the unit-tests working in Travis CI 
(they work on my machine) I'll try to find my stash of clever and take a more extended 
look.

If something is possible, I will probably write some code to make sure that the
common code matches what is currently in Character.scala and re2s.

LeeTibbert replied:
Given the topic here about reducing binary code size, can I ask an implementation
question.

Normally I would consider scalanative.re2s and its unicode data to be more
primitive that java.lang.Character.scala, so I would have the latter call into the former.

I am guessing that Character.scala gets brought into even the simplest HelloWorld 
program.  I am also guessing that re2s will be brought in relatively infrequently.
Do I need to put the Unicode data in a separate `object` in re2s so that only it
is brought in by Character.scala, not all  or too much of re2s?  

Since correctness & time-available are competing, If Character is always being
brought in, I could reverse things and have re2s, if linked, call into Character. 
A bit weird, but i could comment it.

densh replied:
We can make a unicode module somewhere in runtime with all of the unicode data, and then both Character and re2s can both call to it.

LeeTibbert replied:
Understood & agreed. Thank you for the `runtime` placement suggestion.

LeeTibbert replied:
Created Issue #1542 as a result of discussions.

ekrich commented:
I was thinking that we could put the transformations from the Unicode database to create the lists. This might be a good project to eventually put in portable-scala. I had this on my to do list but it is so far down I have made no progress. I have the very beginnings some where on my machine but maybe I can resurrect it and push it to Github.

LeeTibbert commented:
@ekrich I'll learn from and consider whatever you post.  If you do not want to take the time
to clean it up for publix exposure in GitHub, we could probably figure out a way for you to 
send me a private no-warranty copy.   At this stage, everything in my part is just study & dreaming.
Thank you.  L.

LeeTibbert commented:
##### Status update: 

 I believe that I am encountering a situation reported as Issue #1008 by @shadaj.
He reported problems running large number of unit-tests.  

The N re2s tests run as an ensemble in my development environment. 

In the Travis CI environment up to N-2 of the tests can run. The tests chosen to reduce
from N to N - 2 run if they are included in the N - 2 and some other test factored out. Sort of like the statistical procedure known as a jackknife.

I suspect an OutOfMemory issue or such in the child process is causing it to die in the void an be
unable to report iback to the test runner. 

I am going to put the unit-test issue in Travis CI on the back burner as long as all the tests
continue to run in my environment.

densh commented:
@LeeTibbert Next step is to delete existing cre2-based regular expressions and replace them with re2s implementation.

LeeTibbert commented:
> @LeeTibbert Next step is to delete existing cre2-based regular expressions and replace them with re2s implementation.

My thinking had been to do that in an immediate follow-on PR to reduce the number of
items in motion at one time.  At some point, someone might want to look at the port
in isolation, to track down flaws in the port or to do it again.

Would you like to do the cre2 removal as part of this PR?  No problem. 
Your call especially since you are the primary reviewer.

LeeTibbert commented:
#### Status: 2019-04-07 17:00 EDT

I have been studying the Unicode implementations in this PR and in Character.scala.  Making
them common is a good idea, but bigger than a breadbox.  The issue is bigger than a breadbox
but probably not insurmountable.  The same information is encoded in the two areas but in
different ways.   It will take me awhile to figure this out and come up with short & intermediate
range solutions.  More as I learn more.

densh commented:
> Would you like to do the cre2 removal as part of this PR? No problem. 

Yep, I think it's best to do it here. Given that it's supposed to be a drop-in replacement. 

> I have been studying the Unicode implementations in this PR and in Character.scala. Making
them common is a good idea, but bigger than a breadbox. The issue is bigger than a breadbox
but probably not insurmountable

We can postpone unification of unicode data until a follow-up PR. It's not a blocker.

LeeTibbert commented:
Status update: 2019-04-15

The latest commit changes java.util.regex to use re2s.  The code is ready for high level review.
There were major changes in j.u.r.Pattern.scala & j.u.r.Matcher.scala.  Reviewers may want to
consider those new files, based on existing work. rather than trying to figure out the git 
differences.

The current code is certainly alpha or even beta-1 quality. With the exception of a hard
to reach bug in a new Java 8 feature (regions), the code delivers results as good as or
better than the cre2 code.  I have _no_idea_ about changes in execution time.

 Since binary size, code complexity, and maintainability are all concerns, I would like to take
a run at getting j.u.r.Matcher to do more delegation.  The current pass was designed to 
follow the existing j.u.r.Matcher code but use the re2s match engine.  I believe more delegaton
would bring benefits, if I can cut through the complexity.

The latest commit lets me publish the solution "in the large' for consideration by a larger
audience while I focus on the Matcher aspect.  That allows me to reduce the number of
files competing for memory in my brain.

regex support is one of those problems which will take all the time and people-power you
can throw at it and still have just one more corner case.  I believe that after the delegation work,
successful or not, we are approaching the Pareto point of diminishing yield.  Economics, not
an argument for publishing bad code.

densh commented on LICENSE.md:
````diff
@@ -1789,3 +1789,49 @@ The original license notice is included below:
  */
 ```
+
+# License notice for  re2s regular expression code
````
Please delete a section called `License notice for marcomaggi/cre2` above. We don't use cre2 so it's not needed any longer.

LeeTibbert replied:
Done.

 I was not sure if it needed to be retained because both the original SN devo
and I looked at and benefited from that code. In my prior practice, that would
have 'tainted' us.  Thank you for the umpires call, which, as always, if final.

densh commented on javalib/src/main/scala/java/util/regex/cre2.scala:
```diff
@@ -1,167 +0,0 @@
-// re2 c wrapper: https://github.com/marcomaggi/cre2
-@link("re2")
-@extern
-object cre2 {
```
Please also delete `nativelib/src/main/resources/optional/re2.{h, cpp}`.

densh replied:
A few more places where c-based re2 is referenced:
1. scripts/scala-native.nix needs to drop `re2` line.
2. scripts/travis_setup.sh we **can't** drop it there as this script used by 0.3.x users as well.
3. docs/lib/javalib.rst needs to drop all lines that contain `cre2`.
4. docs/user/setup.rst needs to drop `re2` in all setup instructions.

LeeTibbert replied:
Done. travis_setup.sh not touched.  Sorry I missed so many places.

LeeTibbert replied:
I have no way to test scala-native.nix. Will Travis CI do that or can you suggest a way?
I would  rest easier knowing I did not inflict breakage on someone else.  

I will test the  *.rst changes on my machine so they get exercised before the
ramp up to release.

densh commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -46,287 +45,410 @@ object Matcher {
+  private var _groupCount = currentPattern.compiled.groupCount()
+
+  // This default differs from JVM but describes actual SN behavior.
+  // That behavior is a bug, awaiting a fix.
```
Please file an issue and refer to it here.

LeeTibbert replied:
Done. Issue #1541.   

I will try to look at this after the current PR and a rest period.
This is all new work, so nothing in 0.3.8 is broken.  There are a number of regex
issues, all in obscure places, which are competing for attention.  After my brain
heals, I need to do a priority sort.

densh commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -46,287 +45,410 @@ object Matcher {
+    // "groups" is an old-style 2-D row-major array, where Array(index) holds
+    // start for group number index and Array(index + 1) holds the
+    // corresponding end. In SN, 1-D arrays are much more efficient, albeit
+    // much less elegant.
```
1D array are only more *size* efficient for large constants such as unicode tables. Otherwise there is no reason not to use 2D arrays for dynamically-sized data if it's easier to understand.

LeeTibbert replied:
Entirely gone in j.u.r.Matcher.scala in 2019-04-16 code with improved delegation.

I verified that re2s.Matcher.scala uses 1-D arrays.

densh commented on javalib/src/main/scala/java/util/regex/Matcher.scala:
```diff
@@ -46,287 +45,410 @@ object Matcher {
-  }
+    if (doGroups) {
+      val m =
+        Pattern.compile("(\\$(\\d)|\\$\\{(\\w*)\\})").matcher(replacement)
```
Can you cache this `Pattern.compile` as a lazy val at top level somewhere?

LeeTibbert replied:
My most recent iteration of Matcher.scala does more delegation and is much simpler than
the iteration you looked at.  I believe the section you commented on no longer exists
in j.u.r.Matcher.scala.

I checked that no similar code exists in re2s.Matcher.scala.  The two are often parallel but
not here.

densh commented on unit-tests/src/test/scala/scala/scalanative/re2s/ExecTest.scala.disabled:
```diff
@@ -0,0 +1,633 @@
+package scala.scalanative
```
Is this test still failing? Can we enable it?

LeeTibbert replied:
ExecTest.scala itself almost certainly wants to stay disabled. At least until we get some
sort of tiered once-a-week not every CI scheme worked out.  Agreed, a back burner
project.  Those tests are, by design, exhaustive and take tens of minutes (20? 30?)
to run. They are useful enough to be visible to future devos in the Repository, rather
than squirreled away in my boneyard.  Your thoughts?

I believe there are other passing tests which I disabled because Travis was failing when
I enabled the full set.   After I fsync my current batch of changes I an go back and
see if I figure out why Travis is complaining.   Any given N - 2 of the tests pass on
Travis, but running the full N fails.

densh replied:
@LeeTibbert Ok, lets keep it disabled for now.

densh commented on javalib/src/main/scala/java/util/regex/Pattern.scala:
```diff
@@ -19,243 +22,116 @@ object Pattern {
-    // Used to quickly look up a Node from a Key.
-    private val map = scala.collection.mutable.HashMap.empty[Key, Node]
-
-    private def selectNode(regex: String, flags: Int): Node = {
```
The code here implements a bit overdesigned LRU cache. It would be good to keep something like that in the new implementation (i.e. cache compiled instances of regex per input string).

LeeTibbert replied:
I believe I removed that code when I converted to re2s.  

As much as I love caches, I convinced myself that the Pattern cache code 
had not much upside and a lot of down side.  

Much as I like caches, this seemed to me like one of those places to go simple and correct
until  real data shows the need.  Would a cache help at all?

To get the re2s transition to happen I decided to drop the code until experience
and data show that a cache is helpful.  The work could be done in a future sprint.

Compiling a pattern is basically running the parser over it and returning the same thing
if it parses. One must balance the cost of a fast parser versus a cache scheme with
memory allocation.  I made the bet that the former will be faster in almost all use cases.
Of course, there will be special cases, such as repeated use of the same pattern,
where i loose the bet.

Now, I am willing to believe that any serious user of regex is going to be caching themselves
and not relying upon the implementation. Rightly, the JVM documentation says nothing
about caching.

Your thoughts?

densh replied:
A number of string methods internally use regex but don't allow user to cache the regular expressions such as `java.lang.String.{replaceAll, replaceFirst, split, matches}`. So caching the compilation of the regex can reduce overhead significantly, especially when performing those operations on small strings.

LeeTibbert replied:
Understood.  

Can this be spun off into its own Issue and not block this PR?  I think it is going to take
some days and some discussion to get regex pattern caching right.  For example,
informed gestimates of cache size (probably > 10 and < 100).  I must consider 
interactions with underlying RE2 caching. On a first run through it looks like straight
delegation will not work, but then again, tell an engineer that something will not work
and 48 hours later they will have a solution.

Lastly, I must make sure that whatever I write does not trigger issue #1091. Easy
enough to avoid as long as the defect is kept in mind.

Write an LRU cache seems like a job interview question, to be done in 5 minutes with
people judging:  guess I would fail that interview.  The simple list is obvious and
non-concurrent and I tend to write industrial strength code, perhaps at the expense
of timeliness.  That is my excuse for the day and I am sticking with it (Locale specific joke!)

Please advise.  Thank you.

ekrich replied:
I would fail that interview too.

densh replied:
It's perfectly fine to file this for later. Lets do that.

LeeTibbert commented:
I committed changes to regex.Matcher.scala which delegate almost all of the work to re2s and
Travis is still all green.  While I can not call the delegation work elegant, it is at least
more presentable than what was there before.

Thank you @ekrich for using the word 'delegation'. It motivated me to evolve my
code.  The first step was getting regex to use re2s at all. I think the current evolution
is better.

That commit also contained responses to review requests and some
changes to have SN regex use the same exceptions and exception messages the JVM.

I still have one small change to do in Matcher.scala  and several day of polish work
writing Issues, making sure I have addressed review comments, and changing the PR
creation comment.  All work that needs to be done by the light of day, not when I am
tired.

Getting closer!

LeeTibbert commented:
FYI:  some build sanity checks:

* I built the .rst files in my private docs directory.  Both files I touch converted well. There seemed to
be some warnings for interop.rst but I did not track them down.  I viewed the resultant .html files
in Mozilla.  My edits seemed to be there and page movement worked well.

* I converted the LICENSE.md I submitted to PDF and viewed it.  The conversion worked and
the resultant file contained my edits.

densh commented on docs/user/setup.rst:
```diff
@@ -86,14 +80,14 @@ installation of Arch Linux.
 
     $ sudo dnf install llvm clang
-    $ sudo dnf install libunwind-devel gc-devel zlib-devel re2-devel # optional
+    $ sudo dnf install libunwind-devel gc-devel zlib-devel # optional
```
We don't need libunwind deps any more either.

LeeTibbert replied:
Done.  Built on my local machine & viewed in Firefox.

I also searched  all the .rst file in the docs tree for the case-insensitive string "unw".
That showed a few references, but nothing that I could see related to the full 
libunwind.

LeeTibbert replied:
As long as I was in setup.html making non re2s related changes, it seemed fair game
for me to change the description of the default garbage collector to 'immix'. I also
provided a hyperlink to the original paper and did some rewording.

All builds & views. Hyperlink from setup.rst to intended section of sbt.rst works.

densh commented on docs/user/setup.rst:
```diff
@@ -86,14 +78,14 @@ installation of Arch Linux.
 
     $ sudo dnf install llvm clang
-    $ sudo dnf install libunwind-devel gc-devel zlib-devel re2-devel # optional
+    $ sudo dnf install gc-devel zlib-devel # both optional
```
You can split this in two lines and then say something like:
```
$ sudo dnf install gc-devel # optional, only if you use boehm gc
$ sudo dnf install gc-devel # optional, only if you use java.util.zip.*
```
IMO this makes it more clear what are these dependencies for. 

It would be good to update the comment in this style everywhere for optional deps in setup instructions.

densh replied:
Lets change this later in another PR, it's not related to re2s.

densh commented:
LGTM, the work done here is more than good enough for a first PR to get in. There are plenty of things that can be improved, but all of them nowhere near as critical as getting rid of C-based RE2 implementation. 

Thanks to @LeeTibbert for doing the last mile of work to get this into the main repo, and thanks to @techaddict and @MasseGuillaume for initial porting effort on scala-native/re2s ! 👏👏👏
